lib/netext/httpext/compression.go

```diff
@@ -0,0 +1,208 @@
+			contentEncoding += ", "
+		}
+		contentEncoding += compressionType.String()
+		var w io.WriteCloser
+		switch compressionType {
+		case CompressionTypeGzip:
+			w = gzip.NewWriter(buf)
+		case CompressionTypeDeflate:
+			w = zlib.NewWriter(buf)
+		case CompressionTypeZstd:
+			w, _ = zstd.NewWriter(buf)
+		case CompressionTypeBr:
+			w = brotli.NewWriter(buf)
+		default:
+			return nil, "", fmt.Errorf("unknown compressionType %s", compressionType)
+		}
+		// we don't close in defer because zlib will write it's checksum again if it closes twice :(
+		var _, err = io.Copy(w, prevBuf)
+		if err != nil {
+			_ = w.Close()
+			return nil, "", err
+		}
+
+		if err = w.Close(); err != nil {
+			return nil, "", err
+		}
+	}
+
+	return buf, contentEncoding, body.Close()
+}
+
+//nolint:gochecknoglobals
+var decompressionErrors = [...]error{
+	zlib.ErrChecksum, zlib.ErrDictionary, zlib.ErrHeader,
+	gzip.ErrChecksum, gzip.ErrHeader,
+	// TODO: handle brotli errors - currently unexported
+	zstd.ErrReservedBlockType, zstd.ErrCompressedSizeTooBig, zstd.ErrBlockTooSmall, zstd.ErrMagicMismatch,
+	zstd.ErrWindowSizeExceeded, zstd.ErrWindowSizeTooSmall, zstd.ErrDecoderSizeExceeded, zstd.ErrUnknownDictionary,
+	zstd.ErrFrameSizeExceeded, zstd.ErrCRCMismatch, zstd.ErrDecoderClosed,
+}
+
+func newDecompressionError(originalErr error) K6Error {
+	return NewK6Error(
+		responseDecompressionErrorCode,
+		fmt.Sprintf("error decompressing response body (%s)", originalErr.Error()),
+		originalErr,
+	)
+}
+
+func wrapDecompressionError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	// TODO: something more optimized? for example, we won't get zstd errors if
+	// we don't use it... maybe the code that builds the decompression readers
+	// could also add an appropriate error-wrapper layer?
+	for _, decErr := range decompressionErrors {
+		if err == decErr {
+			return newDecompressionError(err)
+		}
+	}
+	if strings.HasPrefix(err.Error(), "brotli: ") { // TODO: submit an upstream patch and fix...
+		return newDecompressionError(err)
+	}
+	return err
+}
+
+func readResponseBody(
+	state *lib.State,
+	respType ResponseType,
+	resp *http.Response,
+	respErr error,
+) (interface{}, error) {
+	if resp == nil || respErr != nil {
+		return nil, respErr
+	}
+
+	if respType == ResponseTypeNone {
+		_, err := io.Copy(ioutil.Discard, resp.Body)
+		_ = resp.Body.Close()
+		if err != nil {
+			respErr = err
+		}
+		return nil, respErr
+	}
+
+	rc := &readCloser{resp.Body}
+	// Ensure that the entire response body is read and closed, e.g. in case of decoding errors
+	defer func(respBody io.ReadCloser) {
+		_, _ = io.Copy(ioutil.Discard, respBody)
+		_ = respBody.Close()
+	}(resp.Body)
+
+	contentEncodings := strings.Split(resp.Header.Get("Content-Encoding"), ",")
+	for i := len(contentEncodings)/2 - 1; i >= 0; i-- {
+		opp := len(contentEncodings) - 1 - i
+		contentEncodings[i], contentEncodings[opp] = contentEncodings[opp], contentEncodings[i]
+	}
```

Comment:
You could use sort.Reverse with sort.StringSlice or just reverse ranging in the for below,
 also a comment explaining why this is done is probably a good idea

Reply:
sort.Reverse is not what we want, since it will sort base on string comparison, not order in list. Reverse range can't be done simply, because we have to check for len(contentEncodings) greater than 0 before do `for i:= len(contentEncodings); i >=0; i--`

Reply:
Instead of reversing the slice, won't it be much simpler to traverse it in reverse order in the `for` loop below?

Reply:
@na-- See my explain above.

I just follow the idiom in go wiki https://github.com/golang/go/wiki/SliceTricks#reversing

Reply:
Yeah, I know the idiom, but it's still difficult to read code that's unnecessary in this case. And if you do the `for` loop like this: `i:= len(contentEncodings)-1; i >=0; i--`, won't that solve the problem?

Reply:
@na-- @MStoykov I will push a fix with a simpler version.